sdks/java/io/contextual-text-io/src/main/java/org/apache/beam/sdk/io/ContextualTextIO/ContextualTextIO.java

```diff
@@ -0,0 +1,570 @@
+              .apply("", View.asMap());
+
+      return linesGroupedByFileAndRange.apply(
+          "assignLineNums",
+          ParDo.of(new assignLineNums(sizesOrdered)).withSideInputs(sizesOrdered));
+    }
+
+    protected static class addFileNameAndRange
+        extends DoFn<LineContext, KV<KV<String, Long>, LineContext>> {
+      @ProcessElement
+      public void processElement(
+          @Element LineContext line, OutputReceiver<KV<KV<String, Long>, LineContext>> out) {
+        out.output(KV.of(KV.of(line.getFile(), line.getRange().getRangeNum()), line));
+      }
+    }
+
+    /** Helper class for computing Number of Lines preceding each Pair of (File, Range) */
+    protected static class computeLinesBeforeEachRange
+        extends DoFn<Integer, KV<KV<String, Long>, Long>> {
+      private final PCollectionView<Map<KV<String, Long>, Long>> sizes;
+
+      public computeLinesBeforeEachRange(PCollectionView<Map<KV<String, Long>, Long>> sizes) {
+        this.sizes = sizes;
+      }
+
+      // Add custom comparator as KV<K, V> is not comparable by default
+      private static class FileRangeComparator<K extends Comparable<K>, V extends Comparable<V>>
+          implements Comparator<KV<K, V>> {
+        @Override
+        public int compare(KV<K, V> a, KV<K, V> b) {
+          if (a.getKey().compareTo(b.getKey()) == 0) {
+            return a.getValue().compareTo(b.getValue());
+          }
+          return a.getKey().compareTo(b.getKey());
+        }
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext p) {
+        // Get the Map Containing the size from side-input
+        Map<KV<String, Long>, Long> sizeMap = p.sideInput(sizes);
+
+        // The FileRange Pair must be sorted
+        SortedMap<KV<String, Long>, Long> sorted = new TreeMap<>(new FileRangeComparator<>());
+
+        // Initialize sorted map with values
+        for (Map.Entry<KV<String, Long>, Long> entry : sizeMap.entrySet()) {
+          sorted.put(entry.getKey(), entry.getValue());
+        }
+
+        // HashMap that tracks lines passed for each file
+        Map<String, Long> pastLines = new HashMap<>();
+
+        // For each (File, Range) Pair, compute the number of lines before it
+        for (Map.Entry entry : sorted.entrySet()) {
+          Long lines = (long) entry.getValue();
+          KV<String, Long> FileRange = (KV<String, Long>) entry.getKey();
+          String file = FileRange.getKey();
+          Long linesBefore = 0L;
+          if (pastLines.containsKey(file)) {
+            linesBefore = pastLines.get(file);
+          }
+          p.output(KV.of(FileRange, linesBefore));
+          pastLines.put(file, linesBefore + lines);
+        }
+      }
+    }
+
+    protected static class assignLineNums
+        extends DoFn<KV<KV<String, Long>, LineContext>, LineContext> {
+      PCollectionView<Map<KV<String, Long>, Long>> sizesOrdered;
```

Comment:
TODO(offline): What does this map represent? Does it have to be ordered?

How about:
fileRangeStartingOffset
fileRangeStartingLine